feat[#3]: 회원가입 작업

hunnypooh opened the pull request:
resolved #3 

## 작업 개요
- 로그인, 로그아웃 제외하고 api 작성. 
- 프로젝트 구조에 맞춰서 작성하고, 컨트롤러 부분만 테스트하는 코드 작성. 

## 작업 사항
- 회원가입(쿠키x), 이메일 중복 검사, 유저 확인 api 작성.
- 쿠키/토큰에서 이메일 가져와서 유저 확인 해야하는데, 아직 그부분 안되서 쿼리 파라미터로 이메일 받아오도록 작성.

## 고민한 점들
### 예외 관련
- 커스텀 exception 만들고, 이걸 핸들링하는 global exception handler로 http 상태코드, 에러 코드, 메세지 보내도록 함. 
- 에러 코드?는 USER_NOT_FOUND_BY_EMAIL 로 하려다가, 회의때 에러 코드?말이 나온거 같아서 임시로 `E001` 같은거로 붙였는데 별로면 USER_NOT_FOUND_BY_EMAIL 이렇게 바꿀게!
- 에러 응답 보낼때, 정해진 에러코드/메세지 말고도 추가 정보 필요할 상황 있을거같아서 CustomException 클래스에 nullable한 additionalInfo 필드 만들어 뒀는데, 아직 에러 응답 보낸거에 추가는 안했음. 얘도 필요할거같으면 넣고, 아니면 빼는 방식으로 수정할게
### 기타
- tdd를 어떻게 할까 생각하다가 일단 컨트롤러만 테스트하는 코드만 작성함. 추가로 테스트 할 레이어가 있으면 추천 부탁!
- repository를 나중에 jpa 안쓰고도 할 수 있게  조금 복잡하게 구성해봄.
- profile, properties 파일은 아직 아예 안 건드림. 나중에 prod, dev, local로 나눠야할거 같은데 지금은 잘 모르겠음ㅠㅠ

hunnypooh committed:
feat: 로그인 UI 작업 [#3]

kimyu0218 commented on backend/src/main/java/com/brick/demo/auth/dto/DuplicateEmailRequestDto.java:
```diff
@@ -0,0 +1,22 @@
+  @Builder
+  DuplicateEmailRequestDto(String email) {
+    this.email = email;
+  }
```
P3
이 DTO에는 필드가 하나 뿐이니까 buidler 사용하면 오버헤드일 것 같아요!
확장 가능성이 없고, 필드가 적은 DTO에는 생성자로 값을 할당해도 좋을 것 같습니다 😆 

그리고 생성자로 값을 할당하고 불변하는 경우에는 `final` 키워드를 활용해보는 건 어떨까요?

hunnypooh replied:
빌더패턴에서의 오버헤드는 생각하지 못했네요! 수정해서 커밋했습니다ㅎㅎ

kimyu0218 commented on backend/src/main/java/com/brick/demo/auth/controller/AuthController.java:
```diff
@@ -0,0 +1,59 @@
+  @PostMapping("/users/duplicate-email")
+  public ResponseEntity<DuplicateEmailResponseDto> duplicateEmail(
+      @RequestBody DuplicateEmailRequestDto dto) {
+    boolean isDuplicate = authService.isExistedEmail(dto);
+    DuplicateEmailResponseDto response = DuplicateEmailResponseDto.builder().
+        duplicateEmail(isDuplicate).
+        build();
```
P2
컨트롤러보다 서비스에 넣는 게 좀 더 자연스러워 보여요!
비즈니스 로직처럼 보이는데 서비스에 작성하는 게 어떨까요?

그리고 `isExistedEmail`이 중복된 이메일 여부를 확인하는 용도니까 함수명을 변경하고 `return authService.isDuplicateEmail(dto)`로 바로 리턴하는 것도 좋아보여요!

hunnypooh replied:
네 좋습니다!

hunnypooh replied:
수정해서 커밋했습니다

kimyu0218 commented on backend/src/main/java/com/brick/demo/auth/controller/AuthController.java:
```diff
@@ -0,0 +1,59 @@
+  @PostMapping("/signup")
+  @ResponseStatus(HttpStatus.CREATED)
+  public void createAccount(@Valid @RequestBody SignUpRequestDto dto) {
+    authService.createAccount(dto);
```
❓여기는 `return`문을 적지 않으셨는데 이유가 있으신가요?

kimyu0218 commented on backend/src/main/java/com/brick/demo/auth/dto/SignUpRequestDto.java:
```diff
@@ -0,0 +1,39 @@
+    this.name = name;
+  }
+
+  public Account toEntity() {
+    return Account.builder()
+        .email(this.email)
+        .password(this.password)
+        .name(this.name)
+        .build();
```
LGTM 👍👍

kimyu0218 commented on backend/src/main/java/com/brick/demo/auth/entity/Account.java:
```diff
@@ -0,0 +1,57 @@
+import java.time.LocalDateTime;
+
+@Entity
+@Table(name = "ACCOUNT")
```
P1
테이블명을 대문자로 작성하신 건가요?
컬럼과 네이밍 방식을 통일하면 좋을 것 같아요! (UPPER CASE vs. snake_case)

hunnypooh replied:
넵 그러면 snake로 통일하겠습니다

kimyu0218 commented on backend/src/main/java/com/brick/demo/auth/entity/Account.java:
```diff
@@ -0,0 +1,57 @@
+    private OAuthProvider oauthProvider;
+
+    @CreatedDate
+    @Column(name = "created_at", nullable = false, updatable = false)
```
P2
DB 설정에서 네이밍 방식을 설정하는 방법이 있을 것 같아요.
NestJS의 경우 다음처럼 네이밍을 통일할 수 있었어요!
자바에서 디폴트로 어떤 네이밍 방식을 적용하는지도 알아보고 필요하면 적용해보시면 좋을 것 같습니다 😄 
```ts
@Module({
  imports: [
    TypeOrmModule.forRootAsync({
      useFactory: () => {
        return {
          type: 'mysql',
          ...
          namingStrategy: new SnakeNamingStrategy(),
          ...
        };
      },
    }),
  ],
})
```

kimyu0218 replied:
💬  @hunnypooh [application.yml](https://github.com/kimyu0218/spring-practice/blob/7074119961f196370a9b42b177a8d6c0c66822df/spring-boot/README/03.Spring_Boot_Spring_Data_JPA.md?plain=1#L50)에서 설정하는 방법을 정리했길래 공유드립니다.
더 이상 `name` 사용하지 않아도 될 것 같아요!

kimyu0218 commented on backend/src/main/java/com/brick/demo/auth/entity/Account.java:
```diff
@@ -0,0 +1,57 @@
+    @Id
+    @Column(name = "id")
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long entityId;
```
❓ `id` 대신 `entityId`를 사용하신 이유가 있나요?
`id`를 사용하면 `@Column` 데코레이터가 필요하지 않을 것 같아서 여쭤봅니다!

kimyu0218 commented on backend/src/main/java/com/brick/demo/auth/repository/AbstractAccountManager.java:
```diff
@@ -0,0 +1,20 @@
+        int ix = myClassName.indexOf("AccountManager");
+        return ix == -1 ? "UNKNOWN" : myClassName.substring(0, ix).toUpperCase(); //AccountManager를 이름에 포함한 구현체 여부.
+    }
+}
```
P4
로깅을 사용하시는 건 좋지만 프로덕션 환경에서는 메모리를 낭비할 수 있어서 꼭 필요한 로깅만 남겨야 해요!
개발 환경에서만 로깅을 활성화하는 방법이 있으면 적용해보시면 좋을 것 같습니다 😄

kimyu0218 replied:
❓그리고 이건 로깅 용도인가요?
repository면 데이터베이스에 접근하는 용도인데 다른 위치가 좀 더 적절할  것 같습니다!

kimyu0218 commented on backend/src/main/java/com/brick/demo/auth/repository/AccountManager.java:
```diff
@@ -0,0 +1,15 @@
+
+import com.brick.demo.auth.entity.Account;
+
+public interface AccountManager {
```
❓ repository 대신 manager를 사용하신 이유가 있으신가요??
repository 밑에 manager가 있어서 조금 어색해요.

kimyu0218 commented on backend/src/main/java/com/brick/demo/auth/repository/JpaAccountManager.java:
```diff
@@ -0,0 +1,41 @@
+
+@Repository
+@NoArgsConstructor
+public class JpaAccountManager extends AbstractAccountManager {
+
+  private EntityManager entityManager;
+
+  @PersistenceContext
+  public void setEntityManager(EntityManager entityManager) {
+    this.entityManager = entityManager;
+  }
+
+  @Override
+  public Account getAccountByEmail(String email) {
+    return entityManager.createQuery("SELECT a FROM Account a WHERE a.email = :email",
+            Account.class)
+        .setParameter("email", email)
+        .getSingleResult();
+  }
+
+  @Override
+  @Transactional
+  public Account save(Account account) {
+    entityManager.persist(account);
+    return account;
+  }
+
+  @Override
+  @Transactional
+  public void update(Account account) {
+    entityManager.merge(account);
+  }
+}
```
P1
Spring Data JPA를 활용해보시는 건 어떤가요?
쿼리나 구현체를 직접 작성할 필요가 없어요!

kimyu0218 commented on backend/src/main/java/com/brick/demo/auth/service/AuthService.java:
```diff
@@ -0,0 +1,40 @@
+  }
+
+  public void createAccount(SignUpRequestDto dto) {
+    accountManager.save(dto.toEntity());
```
P2
리턴값이 있으면 테스트하기 편리해요!
`save`가 아무런 값을 리턴하지 않는다면 `return true`를 통해 메소드가 제대로 수행되었는지 확인할 수 있습니다!
예외가 발생하면 `true`를 리턴하지 않고, 정상적으로 실행되면 `true`를 리턴할 거에요 😄

만약에 `save` 함수에 리턴값이 존재한다면 해당 리턴값을 검증하여 제대로 수행되었는지 검증할 수 있을 것 같습니다!

kimyu0218 commented on backend/src/main/java/com/brick/demo/config/ApplicationConfig.java:
```diff
@@ -0,0 +1,20 @@
+import org.springframework.transaction.annotation.EnableTransactionManagement;
+
+@Configuration
+@EntityScan("auth.entity")
```
P2
엔티티가 생길 때마다 추가하는 건 유지보수가 많이 들 것 같아요.
아래 예시처럼 `basePackages`를 지정하면 좋을 것 같습니다! (spEL 적용해보시면 동작할 것 같아요!)
```java
@EntityScan(basePackages = "hello.hello_spring.domain")
```

kimyu0218 commented on backend/src/main/java/com/brick/demo/auth/entity/OAuthProvider.java:
```diff
@@ -0,0 +1,5 @@
+
+public enum OAuthProvider {
+    KAKAO, GOOGLE
+}
```
❓테이블이 별도로 존재하는 건가요? 
어플리케이션 내에서 `enum`으로만 활용된다면 `Account`에 enum을 정의하시면 더 적합할 것 같습니다.

kimyu0218 reviewed (commented):
양이 많아서 테스트 부분은 리뷰하지 못했습니다 😢
나중에 위의 내용들 수정되면 같이 검토해보겠습니다!! 

💬 Spring Data JPA 부분은 꼭 적용하면 좋을 것 같습니다!
💬 PR 제목에 UI가 있어서 프론트엔드 작업 같은데 다른 제목으로 설정하시는 건 어떤가요?
💬 궁금하신 점은 대댓글로 달아주세요!!

hunnypooh committed:
refactor: isExistedEmail 수정 [#3]
- 이름을 isDuplicatedEmail로 수정.
- controller에 있던것을 service로 옮김.
- 기존 이메일 중복 여부를 판단하는 두가지 경우를 하나의 테스트에 작성했는데, 두개의 다른 테스트로 분리.

hunnypooh committed:
refactor: dto에서 빌더 제거 [#3]
- DuplicateEmailRequestDto, DuplicateEmailResponseDto에서 builder 삭제.
- 테스트 분리하면서 verify에서 두번 호출한 것을 수정했어야 하는데 안했어서 수정.

hunnypooh committed:
refactor: dto에서 빌더 제거 [#3]
- DuplicateEmailRequestDto, DuplicateEmailResponseDto에서 builder 삭제.
- 테스트 분리하면서 verify에서 두번 호출한 것을 수정했어야 하는데 안했어서 수정.

hunnypooh committed:
chore: db 파일 gitignore에 추가

hunnypooh committed:
feat: h2 추가, profile 별로 설정 분리 [#3]

- yml 파일이 properties 보다 가독성이 좋아서 yml로 작성.
- default로 local을 띄우게 했고, 프로덕션 환경에서는 --spring.profiles.active=prod 인수로 실행하면 prod으로 돌아간다고 함.
- 어떤 profile에서 실행 중인지 가독성을 높이기 위해 배너도 profile에 따라서 다르게 띄우도록 설정.

- h2 사용을 위해 의존성 추가.
- http://localhost:8080/h2-console 로 입력하면 db 확인 가능.
- WebSecurityConfig는 웹으로 h2-console 보기 위해서 관련된것만 작성.

hunnypooh committed:
feat: 로그인,로그아웃 [#3]

- 패스워드는 암호화해서 저장.
- 로그인 성공시 쿠키의 'email' 에 사용자 이메일 저장.
- 로그아웃시 쿠키 제거.
- 로그인 성공시 일단 아래처럼 저장하되, 값은 이메일로 전달함.
{
  "token": "string"
}

- 테스트도 쿠키에 이메일을 저장하는 방식을 테스트하도록 작성 (모두 통과해야함)

그외 자잘한 변경사항
- jwt secret 키도 yml에 저장해야해서 application-private.yml로 파일 이름 변경
- api 날리면서 entity, JpaAccountManager (리포지토리) 일부 수정
- build.gradle에 jwt 의존성 추가(아직 사용x)

hunnypooh committed:
feat: 로그인,로그아웃 [#3]

- 로그인시 access, refresh 토큰 발급. 로그인 완료시 Bearer, accessToken, accessToken 만료일 보냄
- 로그인한 상태에서 재 로그인시(메모리의 access token 존재 여부로 판단) exception 보냄.
- 로그아웃 상태에서 다시 로그아웃시 exception 보냄.
- <-근데 이거 두개 원래 401 에러 처리하려고 필터 건드리다가 GlobalExceptionHandler가 안먹어짐 ㅠㅠ

- 로그아웃시 access, refresh 토큰 삭제.

hunnypooh committed:
move: backend 폴더 안에 있는거 밖으로 옮기고 backend 폴더 삭제 [#3]

hunnypooh committed:
move: 도커파일, 워크플로우 옮김 [#3]

hunnypooh committed:
Merge branch 'team-brick-organization:dev' into feat/auth
